src/test/regress/sql/shared_scan.sql

```diff
@@ -137,3 +137,78 @@ select case when ten < 5 then ten else ten * 2 end, count(distinct two), count(d
 select gp_inject_fault('inject_many_fds_for_shareinputscan', 'reset', dbid) from gp_segment_configuration where role = 'p' and content = 0;
 
 \! rm -rf /tmp/_gpdb_fault_inject_tmp_dir/
+
+-- To be able to pass this test, Shared Scan's Material node should be marked
+-- as cross-slice. Previously, we processed all subplans separately from main
+-- plan, which caused Material node not marked as cross-slice (upper main plan's
+-- slice1 motion was ignored in processing).
+-- No error like "ERROR:  cannot execute inactive Motion (nodeMotion.c:264)"
+-- should be shown.
+create table t1 (a int, b int, c int) distributed by (a);
+explain (costs off)
+with cte1 as (
+  select max(c) as c from t1
+), 
+cte2 as (
+  select d as c
+  from generate_series(
+    (select c from cte1) - 4,
+    (select c from cte1), 1) d
+)
+select l.c from cte2 l
+left join t1 u on l.c = u.c;
+
+with cte1 as (
+  select max(c) as c from t1
+), 
+cte2 as (
+  select d as c
+  from generate_series(
+    (select c from cte1) - 4,
+    (select c from cte1), 1) d
+)
+select l.c from cte2 l
+left join t1 u on l.c = u.c;
+
+-- This case shows wrong count(*) result on pre-patched version.
+-- Shared Scan consumer from slice1 not marked as cross-slice and used not
+-- initialized tuple store. We got '1' as the result of the query - only
+-- Shared Scan from slice2 (the part below UNION) executed correctly.
+-- From now, cross-slice interaction detection fixed for subplans and we have
+-- stable '100' as a result.
```

Comment:
This test case is flaky on unpatched version `PostgreSQL 9.4.26 (Greenplum Database 6.20.1_arenadata33 build dev)`:

```
--- \/home\/givi\/src\/gpdb6\/src\/test\/regress\/expected\/shared_scan_optimizer\.out  2022-06-17 10:00:40.097407850 +0000
+++ \/home\/givi\/src\/gpdb6\/src\/test\/regress\/results\/shared_scan\.out     2022-06-17 10:00:40.107407885 +0000
@@ -356,10 +356,7 @@
 )
 select l.c from cte2 l
 left join t1 u on l.c = u.c;
- c 
----
-(0 rows)
-
+ERROR:  cannot execute inactive Motion (nodeMotion.c:264) (nodeMotion.c:264)
 -- This case shows wrong count(*) result on pre-patched version.
 -- Shared Scan consumer from slice1 not marked as cross-slice and used not
 -- initialized tuple store. We got '1' as the result of the query - only

======================================================================
```

Reply:
My bad. It was a stupid question.